### PR TITLE
Hashtree

### PR DESCRIPTION
**Pull request by Licini**
Adding `HashTree` to compare changes on data structrues. For example:
```python
from compas.datastructures import HashTree

print("\nCOMPARE DICTS:")
tree1 = HashTree.from_dict({"a": {"b": 1, "c": 3}, "d": [1, 2, 3], "e": 2})
tree2 = HashTree.from_dict({"a": {"b": 1, "c": 2}, "d": [1, 2, 3], "f": 2})

tree1.print_hierarchy()
tree2.print_hierarchy()

tree2.print_diff(tree1)

print("\nCOMPARE MESH CHANGE:")

from compas.datastructures import Mesh

mesh = Mesh.from_polyhedron(4)
tree1 = HashTree.from_object(mesh)
mesh.vertex_attribute(0, "x", 1.0)
tree2 = HashTree.from_object(mesh)
tree1.print_hierarchy()
tree2.print_hierarchy()
tree2.print_diff(tree1)
```

```
COMPARE DICTS:
└── ROOT @ 4cd56
    ├── .a @ c16fd
    │   ├── .b:1 @ c9b55
    │   └── .c:3 @ 518d4
    ├── .d:[1, 2, 3] @ 9be3a
    └── .e:2 @ 68355
└── ROOT @ fbe39
    ├── .a @ c2022
    │   ├── .b:1 @ c9b55
    │   └── .c:2 @ e3365
    ├── .d:[1, 2, 3] @ 9be3a
    └── .f:2 @ 93861
Added:
{'path': '.f', 'value': 2}
Removed:
{'path': '.e', 'value': 2}
Modified:
{'path': '.a.c', 'old': 3, 'new': 2}

COMPARE MESH CHANGE:
└── ROOT @ 95d3c
    ├── .attributes @ 79a1d
    │   └── .name:Mesh @ 7e7b9
    ├── .dva @ c54cf
    │   ├── .x:0.0 @ 5bc2d
    │   ├── .y:0.0 @ 1704b
    │   └── .z:0.0 @ 6199e
    ├── .dea @ c1cd7
    ├── .dfa @ 977d2
    ├── .vertex @ 631b6
    │   ├── .0 @ 269b3
    │   │   ├── .x:-0.8164965809277261 @ 2a2bf
    │   │   ├── .y:0.0 @ 1704b
    │   │   └── .z:-0.5773502691896258 @ 6244a
    │   ├── .1 @ 70091
    │   │   ├── .x:0.0 @ 5bc2d
    │   │   ├── .y:-0.8164965809277261 @ dcdd6
    │   │   └── .z:0.5773502691896258 @ 2c4c0
    │   ├── .2 @ 71776
    │   │   ├── .x:0.8164965809277261 @ 38e43
    │   │   ├── .y:0.0 @ 1704b
    │   │   └── .z:-0.5773502691896258 @ 6244a
    │   └── .3 @ 9c4f5
    │       ├── .x:0.0 @ 5bc2d
    │       ├── .y:0.8164965809277261 @ 477c9
    │       └── .z:0.5773502691896258 @ 2c4c0
    ├── .face @ 8ed1b
    │   ├── .0:[0, 1, 2] @ 8bb03
    │   ├── .1:[0, 3, 1] @ 15099
    │   ├── .2:[0, 2, 3] @ 96bc4
    │   └── .3:[1, 3, 2] @ d242c
    ├── .facedata @ c21a1
    │   ├── .0 @ a8138
    │   ├── .1 @ 80b33
    │   ├── .2 @ 35d79
    │   └── .3 @ 26f09
    ├── .edgedata @ 9a637
    ├── .max_vertex:3 @ 01770
    └── .max_face:3 @ 77792
└── ROOT @ 0b3b4
    ├── .attributes @ 79a1d
    │   └── .name:Mesh @ 7e7b9
    ├── .dva @ c54cf
    │   ├── .x:0.0 @ 5bc2d
    │   ├── .y:0.0 @ 1704b
    │   └── .z:0.0 @ 6199e
    ├── .dea @ c1cd7
    ├── .dfa @ 977d2
    ├── .vertex @ 24c72
    │   ├── .0 @ bc868
    │   │   ├── .x:1.0 @ 614ef
    │   │   ├── .y:0.0 @ 1704b
    │   │   └── .z:-0.5773502691896258 @ 6244a
    │   ├── .1 @ 70091
    │   │   ├── .x:0.0 @ 5bc2d
    │   │   ├── .y:-0.8164965809277261 @ dcdd6
    │   │   └── .z:0.5773502691896258 @ 2c4c0
    │   ├── .2 @ 71776
    │   │   ├── .x:0.8164965809277261 @ 38e43
    │   │   ├── .y:0.0 @ 1704b
    │   │   └── .z:-0.5773502691896258 @ 6244a
    │   └── .3 @ 9c4f5
    │       ├── .x:0.0 @ 5bc2d
    │       ├── .y:0.8164965809277261 @ 477c9
    │       └── .z:0.5773502691896258 @ 2c4c0
    ├── .face @ 8ed1b
    │   ├── .0:[0, 1, 2] @ 8bb03
    │   ├── .1:[0, 3, 1] @ 15099
    │   ├── .2:[0, 2, 3] @ 96bc4
    │   └── .3:[1, 3, 2] @ d242c
    ├── .facedata @ c21a1
    │   ├── .0 @ a8138
    │   ├── .1 @ 80b33
    │   ├── .2 @ 35d79
    │   └── .3 @ 26f09
    ├── .edgedata @ 9a637
    ├── .max_vertex:3 @ 01770
    └── .max_face:3 @ 77792
Added:
Removed:
Modified:
{'path': '.vertex.0.x', 'old': -0.8164965809277261, 'new': 1.0}
```

### CODE REVIEW

**Comment by gonzalocasas** [comment]
This is really cool! I didn't have time to review it in detail, and I also don't recall exactly how merkle trees are supposed to work, but it's a nice idea, I will go through it probably early next year!

**Comment by Licini** [comment]
> This is really cool! I didn't have time to review it in detail, and I also don't recall exactly how merkle trees are supposed to work, but it's a nice idea, I will go through it probably early next year!

No hurries! It's just put up there for discussion. Have a great holiday! : )

File: src/compas/datastructures/tree/hashtree.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+        return node
+
+
+class HashTree(Tree):
+    """A Hash tree (or Merkle tree) is a tree in which every leaf node is labelled with the cryptographic hash of a data block
+    and every non-leaf node is labelled with the hash of the labels of its child nodes.
+    Hash trees allow efficient and secure verification of the contents of large data structures.
+    They can also be used to compare different versions(states) of the same data structure for changes.
+
```
Comment (tomvanmele) [inline src/compas/datastructures/tree/hashtree.py]: can we split this into a one-liner and then a longer description?

File: src/compas/datastructures/tree/hashtree.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    Attributes
+    ----------
+    signatures : dict[str, str]
+        The SHA256 signatures of the nodes in the tree. The keys are the absolute paths of the nodes, the values are the signatures.
+
+
```
Comment (tomvanmele) [inline src/compas/datastructures/tree/hashtree.py]: remove

File: src/compas/datastructures/tree/hashtree.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        +-- .f:2 @ 93861
+    >>> tree2.print_diff(tree1)
+    Added:
+    {'path': '.f', 'value': 2}
+    Removed:
+    {'path': '.e', 'value': 2}
+    Modified:
+    {'path': '.a.c', 'old': 3, 'new': 2}
+
+
```
Comment (tomvanmele) [inline src/compas/datastructures/tree/hashtree.py]: remove

File: src/compas/datastructures/tree/hashtree.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    """
+
+    def __init__(self):
+        super().__init__()
```
Comment (tomvanmele) [inline src/compas/datastructures/tree/hashtree.py]: kwargs?

**Review by tomvanmele — COMMENTED** [review]
would be good to add a few examples and especially tests

**Comment by tomvanmele** [comment]
@Licini any progress here?

**Comment by jf---** [comment]
@Licini impressive contribution!

**Comment by Licini** [comment]
@tomvanmele could you take another look? And if adding an example, where should it go? tutorial.basics or tutorial.advanced?

**Review by tomvanmele — COMMENTED** [review]
seems to LGTM, but is this meant to be serialisable? `__data__` seems missing then...

**Review by tomvanmele — COMMENTED** [review]
@Licini looks good, but would be even better if you add a (minimal) corresponding page in the docs related to this new data structure :)

**Comment by tomvanmele** [comment]
> seems to LGTM, but is this meant to be serialisable? `__data__` seems missing then...

@Licini ping

**Comment by Licini** [comment]
> @Licini looks good, but would be even better if you add a (minimal) corresponding page in the docs related to this new data structure :)

Hey sorry totally forgot about this one for a while..... Added a tutorial now!

**Review by tomvanmele — APPROVED** [review]

**Comment by jf---** [comment]
@tomvanmele this is a considerable contribution -- what's holding back from merging?

**Comment by tomvanmele** [comment]
@jf--- it was merged yesterday

**Comment by jf---** [comment]
🤦‍♂️

**Pull request merged**